x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvPercentile.java

```diff
@@ -0,0 +1,390 @@
+        } else {
+            assert lowerIndex >= 0 && upperIndex < valueCount;
+            var lowerValue = values[lowerIndex];
+            var upperValue = values[upperIndex];
+            var difference = (long) upperValue - lowerValue;
+            var percentileValue = lowerValue + (int) (fraction * difference);
+            builder.appendInt(percentileValue);
+        }
+    }
+
+    private static void processLongs(LongBlock.Builder builder, int position, LongBlock valuesBlock, double percentile) {
+        int valueCount = valuesBlock.getValueCount(position);
+        int firstValueIndex = valuesBlock.getFirstValueIndex(position);
+
+        if (valueCount == 0 || percentile < 0 || percentile > 100) {
+            builder.appendNull();
+            return;
+        }
+
+        if (valueCount == 1) {
+            builder.appendLong(valuesBlock.getLong(firstValueIndex));
+            return;
+        }
+
+        var p = percentile / 100.0;
+        var index = p * (valueCount - 1);
+        var lowerIndex = (int) index;
+        var upperIndex = lowerIndex + 1;
+        var fraction = index - lowerIndex;
+
+        if (valuesBlock.mvSortedAscending()) {
+            if (percentile == 0) {
+                builder.appendLong(valuesBlock.getLong(0));
+            } else if (percentile == 100) {
+                builder.appendLong(valuesBlock.getLong(valueCount - 1));
+            } else {
+                assert lowerIndex >= 0 && upperIndex < valueCount;
+                var result = calculateLongPercentile(fraction, valuesBlock.getLong(lowerIndex), valuesBlock.getLong(upperIndex));
+                builder.appendLong(result);
+            }
+        }
+
+        var values = new long[valueCount];
+
+        for (int i = 0; i < valueCount; i++) {
+            values[i] = valuesBlock.getLong(firstValueIndex + i);
+        }
+
+        Arrays.sort(values);
+
+        if (percentile == 0) {
+            builder.appendLong(values[0]);
+        } else if (percentile == 100) {
+            builder.appendLong(values[valueCount - 1]);
+        } else {
+            assert lowerIndex >= 0 && upperIndex < valueCount;
+            var result = calculateLongPercentile(fraction, values[lowerIndex], values[upperIndex]);
+            builder.appendLong(result);
+        }
+    }
+
+    /**
+     * Calculates a percentile for a long avoiding overflows and double precision issues.
+     * <p>
+     *     To do that, if the values are over the limit of the representable double integers,
+     *     it uses instead BigDecimals for the calculations.
+     * </p>
+     */
+    private static long calculateLongPercentile(double fraction, long lowerValue, long upperValue) {
+        if (upperValue < MAX_SAFE_LONG_DOUBLE && lowerValue > -MAX_SAFE_LONG_DOUBLE) {
+            var difference = upperValue - lowerValue;
+            return lowerValue + (long) (fraction * difference);
+        }
+
+        var lowerValueBigDecimal = new BigDecimal(lowerValue);
```

Comment:
To avoid overflowing or having precision issues on large longs, I'm operating with BigDecimals instead.

I have doubts about this. Do we prefer speed or precision for big longs?